File: app/src/main/kotlin/com/devhyesun/kolinsample/ui/repository/RepositoryActivity.java
package com.devhyesun.kolinsample.ui.repository;

import android.support.constraint.ConstraintLayout;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.View;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.devhyesun.kolinsample.R;
import com.devhyesun.kolinsample.api.GithubApi;
import com.devhyesun.kolinsample.api.GithubApiProvider;
import com.devhyesun.kolinsample.api.model.GithubRepo;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class RepositoryActivity extends AppCompatActivity {

    public static final String KEY_USER_LOGIN = "user_login";
    public static final String KEY_REPO_NAME = "repo_name";

    ConstraintLayout clContent;
    ImageView ivProfile;
    TextView tvName;
    TextView tvStars;
    TextView tvDescription;
    TextView tvLanguage;
    TextView tvLastUpdate;
    ProgressBar progressBar;
    TextView tvMessage;

    GithubApi api;
    Call<GithubRepo> githubRepoCall;

    SimpleDateFormat dateFormatInResponse = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssX", Locale.getDefault());
    SimpleDateFormat dateFormatToShow = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.getDefault());

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.atv_repository);

        clContent = findViewById(R.id.cl_repository_content);
        ivProfile = findViewById(R.id.iv_repository_profile);
        tvName = findViewById(R.id.tv_repository_name);
        tvStars = findViewById(R.id.tv_repository_star);
        tvDescription = findViewById(R.id.tv_repository_description);
        tvLanguage = findViewById(R.id.tv_repository_language);
        tvLastUpdate = findViewById(R.id.tv_repository_last_update);
        progressBar = findViewById(R.id.pb_repository);
        tvMessage = findViewById(R.id.tv_repository_message);

        api = GithubApiProvider.provideGithubApi(this);

        String login = getIntent().getStringExtra(KEY_USER_LOGIN);
        if(login == null) {
            throw new IllegalArgumentException("No login info exists int extras");
        }

        String repo = getIntent().getStringExtra(KEY_REPO_NAME);
        if(repo == null) {
            throw new IllegalArgumentException("No repo info exists in extras");
        }

        showRepositoryInfo(login, repo);
    }

    private void showRepositoryInfo(String login, String repoName) {
        showProgress();

        githubRepoCall = api.getRepository(login, repoName);
        githubRepoCall.enqueue(new Callback<GithubRepo>() {
            @Override
            public void onResponse(Call<GithubRepo> call, Response<GithubRepo> response) {
                hideProgress(true);

                GithubRepo githubRepo = response.body();
                if(response.isSuccessful() && githubRepo != null) {
                    Glide.with(RepositoryActivity.this)
                            .load(githubRepo.owner.avatarUrl)
                            .into(ivProfile);

                    tvName.setText(githubRepo.fullName);
                    tvStars.setText(getResources().getQuantityString(R.plurals.star, githubRepo.stars, githubRepo.stars));

                    if(githubRepo.description == null) {
                        tvDescription.setText(getString(R.string.no_description_provided));
                    } else {
                        tvDescription.setText(githubRepo.description);
                    }

                    if(githubRepo.language == null) {
                        tvLanguage.setText(getString(R.string.no_language_specified));
                    } else {
                        tvLanguage.setText(githubRepo.language);
                    }

                    try {
                        Date lastUpdate = dateFormatInResponse.parse(githubRepo.updateAt);
                        tvLastUpdate.setText(dateFormatToShow.format(lastUpdate));
                    } catch (ParseException e) {
                        tvLastUpdate.setText(getString(R.string.unknown));
                    }
                } else {
                    showError("Not successful : " + response.message());
                }
            }

            @Override
            public void onFailure(Call<GithubRepo> call, Throwable t) {
                hideProgress(false);
                showError(t.getMessage());
            }
        });
    }

    private void showProgress() {
        clContent.setVisibility(View.GONE);
        progressBar.setVisibility(View.VISIBLE);
    }

    private void hideProgress(boolean isSucceed) {
        clContent.setVisibility(isSucceed ? View.VISIBLE : View.GONE);
        progressBar.setVisibility(View.GONE);
    }

    private void showError(String message) {
        tvMessage.setText(message);
        tvMessage.setVisibility(View.VISIBLE);
    }
}

File: app/src/main/kotlin/com/devhyesun/kolinsample/ui/repository/RepositoryActivity.kt
package com.devhyesun.kolinsample.ui.repository

import android.arch.lifecycle.ViewModelProviders
import android.os.Bundle
import android.view.View

import com.bumptech.glide.Glide
import com.devhyesun.kolinsample.rx.AutoClearedDisposable
import com.devhyesun.kolinsample.R
import com.devhyesun.kolinsample.extensions.plusAssign
import dagger.android.support.DaggerAppCompatActivity
import io.reactivex.android.schedulers.AndroidSchedulers
import kotlinx.android.synthetic.main.atv_repository.*

import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.Locale
import javax.inject.Inject

class RepositoryActivity : DaggerAppCompatActivity() {
    companion object {

        const val KEY_USER_LOGIN = "user_login"
        const val KEY_REPO_NAME = "repo_name"
    }

    @Inject lateinit var viewModelFactory: RepositoryViewModelFactory

    private val disposables = AutoClearedDisposable(this)
    private val viewDisposables =
        AutoClearedDisposable(lifecycleOwner = this, alwaysClearOnStop = false)

    lateinit var viewModel: RepositoryViewModel

    private val dateFormatInResponse = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssX", Locale.getDefault())
    private val dateFormatToShow = SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.getDefault())

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.atv_repository)

        viewModel = ViewModelProviders.of(this, viewModelFactory)[RepositoryViewModel::class.java]

        lifecycle += disposables
        lifecycle += viewDisposables

        viewDisposables += viewModel.repository
            .filter { !it.isEmpty }
            .map { it.value }
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe { repository ->
                Glide.with(this@RepositoryActivity)
                    .load(repository.owner.avatarUrl)
                    .into(iv_repository_profile)
                tv_repository_name.text = repository.fullName
                tv_repository_star.text = resources.getQuantityString(R.plurals.star, repository.stars, repository.stars)
                if(repository.description == null) {
                    tv_repository_description.setText(R.string.no_description_provided)
                } else {
                    tv_repository_description.text = repository.description
                }
                if(repository.language == null) {
                    tv_repository_language.setText(R.string.no_language_specified)
                } else {
                    tv_repository_language.text = repository.language
                }

                try {
                    val lastUpdate = dateFormatInResponse.parse(repository.updateAt)
                    tv_repository_last_update.text = dateFormatToShow.format(lastUpdate)
                } catch (e: ParseException) {
                    tv_repository_last_update.text = getString(R.string.unknown)
                }
            }

        viewDisposables += viewModel.message
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe { message -> showError(message) }

        viewDisposables += viewModel.isContentVisible
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe { visible -> setContentVisible(visible) }

        viewDisposables += viewModel.isLoading
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe { isLoading ->
                if(isLoading) {
                    showProgress()
                } else {
                    hideProgress()
                }
            }

        val login =
            intent.getStringExtra(KEY_USER_LOGIN) ?: throw IllegalArgumentException("No login info exists int extras")

        val repo =
            intent.getStringExtra(KEY_REPO_NAME) ?: throw IllegalArgumentException("No repo info exists in extras")

        disposables += viewModel.requestRepositoryInfo(login, repo)

    }

    private fun showProgress() {
        cl_repository_content.visibility = View.GONE
        pb_repository.visibility = View.VISIBLE
    }

    private fun hideProgress() {
        pb_repository.visibility = View.GONE
    }

    private fun setContentVisible(show: Boolean) {
        cl_repository_content.visibility = if(show) View.VISIBLE else View.GONE
    }

    private fun showError(message: String?) {
        with(tv_repository_message) {
            text = message ?: "Unexpected error."
            visibility = View.VISIBLE
        }
    }
}
